tests/DCPS/XTypes/XTypes.cpp

```diff
@@ -0,0 +1,668 @@
+{
+  MutableUnionDataWriter_var typed_dw = MutableUnionDataWriter::_narrow(dw);
+
+  MutableUnion mu;
+  mu.key(key_value);
+  typed_dw->write(mu, HANDLE_NIL);
+  if (verbose) {
+    ACE_DEBUG((LM_DEBUG, "writer: MutableUnion\n"));
+  }
+}
+
+
+void write_modified_mutable_union(const DataWriter_var& dw)
+{
+  ModifiedMutableUnionDataWriter_var typed_dw = ModifiedMutableUnionDataWriter::_narrow(dw);
+
+  ModifiedMutableUnion mmu;
+  mmu.key(key_value);
+  typed_dw->write(mmu, HANDLE_NIL);
+  if (verbose) {
+    ACE_DEBUG((LM_DEBUG, "writer: ModifiedMutableUnion\n"));
+  }
+}
+
+
+void write_trim64_struct(const DataWriter_var& dw)
+{
+  Trim64StructDataWriter_var typed_dw = Trim64StructDataWriter::_narrow(dw);
+
+  Trim64Struct tcs;
+  tcs.trim_string = STRING_26.c_str();
+  typed_dw->write(tcs, HANDLE_NIL);
+  if (verbose) {
+    ACE_DEBUG((LM_DEBUG, "writer: Trim64Struct\n"));
+  }
+}
+
+
+void write_appendable_struct_with_dependency(const DataWriter_var& dw)
+{
+  AppendableStructWithDependencyDataWriter_var typed_dw = AppendableStructWithDependencyDataWriter::_narrow(dw);
+
+  AppendableStructWithDependency as;
+  as.key = key_value;
+  as.additional_nested_struct.additional_field = NESTED_STRUCT_AF;
+  typed_dw->write(as, HANDLE_NIL);
+  if (verbose) {
+    ACE_DEBUG((LM_DEBUG, "writer: AppendableStructWithDependency\n"));
+  }
+}
+
+template<typename T>
+void get_topic(T ts, const DomainParticipant_var dp, const std::string& topic_name,
+  Topic_var& topic, const std::string& registered_type_name)
+{
+  ts->register_type(dp, registered_type_name.c_str());
+  CORBA::String_var type_name = (registered_type_name.empty() ? ts->get_type_name() : registered_type_name.c_str());
+  topic = dp->create_topic(topic_name.c_str(), type_name,
+    TOPIC_QOS_DEFAULT, 0, DEFAULT_STATUS_MASK);
+}
+
+
+bool check_inconsistent_topic_status(Topic_var topic)
+{
+  DDS::InconsistentTopicStatus status;
+  DDS::ReturnCode_t retcode;
+
+  retcode = topic->get_inconsistent_topic_status(status);
+  if (retcode != DDS::RETCODE_OK) {
+    ACE_ERROR((LM_ERROR, "ERROR: get_inconsistent_topic_status failed\n"));
+    return false;
+  } else if (status.total_count != (expect_to_fail ? 1 : 0)) {
+    ACE_ERROR((LM_ERROR, "ERROR: inconsistent topic count is %d\n", status.total_count));
+    return false;
+  }
+  return true;
+}
+
+
+int ACE_TMAIN(int argc, ACE_TCHAR* argv[])
+{
+  DomainParticipantFactory_var dpf = TheParticipantFactoryWithArgs(argc, argv);
+  DomainParticipant_var dp = dpf->create_participant(23,
+    PARTICIPANT_QOS_DEFAULT, 0, DEFAULT_STATUS_MASK);
+
+  bool writer = false;
+  bool reader = false;
+  std::string type;
+  std::string registered_type_name = "";
+
+  for (int i = 1; i < argc; ++i) {
+    ACE_TString arg(argv[i]);
+    if (arg == ACE_TEXT("--verbose")) {
+      verbose = true;
+    } else if (arg == ACE_TEXT("--writer")) {
+      writer = true;
+    } else if (arg == ACE_TEXT("--reader")) {
+      reader = true;
```

Comment:
I think the test is designed to create either a writer or a reader, but not both, for each process, right? If so, maybe check if the user enters both `--writer` and `--reader` options, and print an error message in that case.